flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/aggregate/over/frame/RowSlidingOverFrame.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.runtime.aggregate.over.frame;
+
+import org.apache.flink.table.dataformat.BaseRow;
+import org.apache.flink.table.generated.GeneratedAggsHandleFunction;
+import org.apache.flink.table.runtime.util.ResettableExternalBuffer;
+import org.apache.flink.table.type.RowType;
+
+/**
+ * The sliding window frame calculates frames with the following SQL form:
+ * ... ROW BETWEEN 1 PRECEDING AND 1 FOLLOWING
+ */
+public class RowSlidingOverFrame extends SlidingOverFrame {
+
+	private final int leftOffset;
+	private final int rightOffset;
+
+	/**
+	 * Index of the first input row with a value greater than the upper bound of the current
+	 * output row.
+	 */
+	private int inputHighIndex = 0;
+
+	/**
+	 * Index of the first input row with a value equal to or greater than the lower bound of the
+	 * current output row.
+	 */
+	private int inputLowIndex = 0;
+
+	public RowSlidingOverFrame(
+			RowType inputType,
+			RowType valueType,
+			GeneratedAggsHandleFunction aggsHandleFunction,
+			int leftOffset,
+			int rightOffset) {
+		super(inputType, valueType, aggsHandleFunction);
+		this.leftOffset = leftOffset;
+		this.rightOffset = rightOffset;
+	}
+
+	@Override
+	public void prepare(ResettableExternalBuffer rows) throws Exception {
+		super.prepare(rows);
+		inputHighIndex = 0;
+		inputLowIndex = 0;
+	}
+
+	@Override
+	public BaseRow process(int index, BaseRow current) throws Exception {
+		boolean bufferUpdated = index == 0;
+
+		// Drop all rows from the buffer for which the input row value is smaller than
+		// the output row lower bound.
+		while (!buffer.isEmpty() && inputLowIndex < index + leftOffset) {
```

Comment:
index - leftOffset?

Reply:
leftOffset is negative value.
I will change it to positive value to better understood.